insights-notification-schemas-java/src/main/resources/schemas/Action-out.json

```diff
@@ -0,0 +1,58 @@
+{
+  "$comment": "This schema defines some additional information for an Action that might be useful",
+  "$schema": "https://json-schema.org/draft-07/schema#",
+  "additionalProperties": false,
```

Comment:
With the use of `"additionalProperties": false` we restrict the chances of combining schemas.
[fge gives a good explanation](https://stackoverflow.com/a/23001194/688379).

So anyway, what we can do here (without removing the `additionalProperties` in both schemas) is to add to this schema all the properties of `Action` (you don't need to copy the details, only the top level names.
i.e. 
```
// In Action-out-json

"properties": {
  "version": {}, // yep, an empty object should be enough
  "id": {},
   // and so on for the top properties.
}
```

Reply:
Thank you so much for that link! That was insightful! Following that link, I saw that there exists the `merge` option which could be applied to the `Action-out` thing... although it requires a schema validator that supports the "draft 5" at least... What do you think?

---

I ended up copying and pasting everything, since we're at it... Because I'm a little bit afraid that someone might send an action with a version `"version": 1.2.3` instead of `"version": "1.2.3"`... It shouldn't happen, but...

Reply:
> EDIT: I just saw that the json-schema-validator implements the "draft 4", so just ignore this comment ￼

According do their docs:

> Supports JSON Schema Draft 3, Draft 4, Draft 6, Draft 7 and Draft 2019-09.

But I'm unsure if the `merge` option is still a thing in the latest drafts.

> EDIT: I just understood why you suggested that... this way it doesn't generate clashes when generating the classes... Nice one!

Yeah, by specifying the `allOF` and the properties, one will validate that they exist, and the other will validate that they are the correct ypes - allowing the properties to exist (because we are preventing the unknown properties).

Might require a bit more of tweaking though. Ultimately I guess we can remove the `additionalProperties: false`.

Reply:
All right! I have provided some tests for the `source` key.  I think you're right, though, maybe we can simply append the `source` key to the `Action.json` schema, since I've done it so that the validator doesn't complain even if the source object is empty.

The only time it complains is when the source object contains any of the "application", "bundle" or "event type" keys, and if they're empty. Although we could change that too.